addon/shield.js

```diff
@@ -0,0 +1,201 @@
+          // increment the connect counter to avoid the endless retry
+          this._connectRetried++;
+          this.reconnectMetadataStore();
+          Cu.reportError(error);
+        });
+    }, 500);
+  },
+
+  loadActivityStream: Task.async(function*() {
+    yield this.setUp();
+    if (this._options.loadReason === "upgrade") {
+      yield this.migrateMetadataStore();
+    }
+    this._metadataStore = new MetadataStore();
+    try {
+      yield this._metadataStore.asyncConnect();
+    } catch (e) {
+      this.reconnectMetadataStore();
+    }
+    this._activityStreamApp = new ActivityStreams(this._metadataStore, this._tabTracker, this._telemetrySender, this._options);
+    // don't override the homepage to have activity stream
+    this._activityStreamApp._setHomePage = () => {};
+    try {
+      this._activityStreamApp.init();
+    } catch (e) {
+      Cu.reportError(e);
+    }
+  }),
+
+  loadTiles: Task.async(function*() {
+    yield this.setUp();
+    this.initializeNewTab();
+    if (!this._placesQueries) {
+      this._placesQueries = {
+        getHistorySize() {return PlacesProvider.links.getHistorySize();},
+        getBookmarksSize() {return PlacesProvider.links.getBookmarksSize();}
+      };
+    }
+    this._tabTracker.init(["about:newtab"], this._placesQueries);
+    this._initializedTiles = true;
+  }),
+
+  initializeNewTab() {
+    let reportEvent = (event, data = {}) => {
+      let payload = Object.assign({event}, data);
+      this._tabTracker.handleUserEvent(payload);
+    };
+
+    // Instrument tile actions, e.g., click
+    this._dlp = Cu.import("resource:///modules/DirectoryLinksProvider.jsm", {}).DirectoryLinksProvider;
+    this._dlp._reportSitesAction = this._dlp.reportSitesAction;
+    this._dlp.reportSitesAction = function(sites, action, action_position) {
+      if (action.search(/^(click|block)$/) === 0) {
+        reportEvent(action.toUpperCase(), {
+          action_position,
+          source: sites[action_position].link.type.toUpperCase()
+        });
+      }
+      return this._reportSitesAction(sites, action, action_position);
+    };
+
+    // Instrument search actions
+    this._tilesPageMod = new PageMod({
+      attachTo: ["existing", "top"],
+      contentScript: `addEventListener("ContentSearchClient", ({detail}) => {
+        if (detail.type === "Search") self.port.emit("search", detail.data)
+      });`,
+      include: "about:newtab",
+      onAttach(worker) {
+        worker.port.on("search", () => reportEvent("SEARCH"));
+      }
+    });
+  },
+
+  isEligible() {
+    this._setUpEligibilityFilters();
+    return this._eligibilityFilters.every(item => item) && !this._doesHaveTestPilot;
+  },
+
+  doesHaveTestPilot() {
+    const testPilotID = "@testpilot";
+    return getAddonByID(testPilotID).then(addon => {
+      this._doesHaveTestPilot = !!addon;
+    });
+  },
+
+  shutdown(reason, variant) {
+    if (this._activityStreamApp && variant === "ActivityStream") {
```

Comment:
This might have been a little bit cleaner if there were two `ActivityStreamVariant` and `TilesVariant` that knew how to `load` and `shutdown`. Right now there's a bunch of internal variables and methods that are only used for one or the other. No need to fix here.